### PR TITLE
Trail componen

### PR DESCRIPTION
**Pull request by telmiuse**
- Script and scene to test Fang shoot
- new function to create gameobjects in scripts
- textures and trails components

### CODE REVIEW

File: Project/Source/Components/ComponentTrail.cpp [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+		verticesPosition[i] = verticesPosition[i + 30];
+	}
+}
+void ComponentTrail::insertVertex(float3 vertex) {
+	verticesPosition[trianglesCreated++] = vertex.x;
+	verticesPosition[trianglesCreated++] = vertex.y;
+	verticesPosition[trianglesCreated++] = vertex.z;
+}
+
+void ComponentTrail::insertTextureCoords() {
```
Comment (JaimeBea) [inline Project/Source/Components/ComponentTrail.cpp:213]: Make sure all the variable and function names are following the code style guide.

File: Project/Source/Components/ComponentTrail.cpp [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	float4x4* proj = &App->camera->GetProjectionMatrix();
+	float4x4* view = &App->camera->GetViewMatrix();
+	//float4x4 MiguelMatrix = float4x4::LookAt(transform->GetGlobalMatrix().RotatePart().Col(2), -frustum->Front(), transform->GetGlobalMatrix().RotatePart().Col(1), float3::unitY);
+	/*float4x4 newModelMatrix = transform->GetGlobalMatrix().LookAt(transform->GetGlobalMatrix().RotatePart().Col(2), -frustum->Front(), transform->GetGlobalMatrix().RotatePart().Col(1), float3::unitY);
+	newModelMatrix = transform->GetGlobalMatrix();
+	float4x4 Final = float4x4::FromTRS(float3::zero, transform->GetGlobalMatrix().RotatePart(), transform->GetGlobalScale());*/
+
+	//-> glUniformMatrix4fv(glGetUniformLocation(program, "model"), 1, GL_TRUE, newModelMatrix.ptr());
+	float3 PolLocalMiuse = transform->GetRotation().ToEulerXYZ();
+	float3 PolMiuse = transform->GetGlobalRotation().ToEulerXYZ();
```
Comment (JaimeBea) [inline Project/Source/Components/ComponentTrail.cpp]: Aha!

File: Project/Source/Components/ComponentTrail.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	float minDistance = 2.0f;
+	/*std::vector<float> particlesPosition;*/
+	float verticesPosition[1500];
+
+	float3 currentPosition;
+	float3 previousPosition;
+	float3 previousVectorUp;
+	float3 currentPositionUp;
+	float3 currentPositionDown;
+	float3 previousPositionUp = float3(0, 0, 0);
```
Comment (JaimeBea) [inline Project/Source/Components/ComponentTrail.h:47]: Make sure all variables are initialized just in case.

File: Project/Source/GameplaySystems.h [hunk 4]
```diff
@@ -490,13 +490,15 @@ namespace Physics {
...
-
 	TESSERACT_ENGINE_API float3 Red();
 	TESSERACT_ENGINE_API float3 White();
 	TESSERACT_ENGINE_API float3 Blue();
 	TESSERACT_ENGINE_API float3 Orange();
 	TESSERACT_ENGINE_API float3 Green();
+<<<<<<< HEAD
+=======
 
+>>>>>>> develop
```
Comment (JaimeBea) [inline Project/Source/GameplaySystems.h]: Merge artifacts.

File: Project/Source/Components/ComponentTrail.cpp [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+#include "Math/float3x3.h"
+#include "Utils/ImGuiUtils.h"
+#include "Math/TransformOps.h"
+#include "imgui.h"
+#include "GL/glew.h"
+#include "debugdraw.h"
+
+#include "Utils/Leaks.h"
+#include <Utils/Logging.h>
+
```
Comment (JaimeBea) [inline Project/Source/Components/ComponentTrail.cpp:28]: Leaks.h should always be the last include.

File: Game/Assets/ybotCasting_Spell.fbx.meta [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        },
+        {
+            "Id": 16107003061144139599,
+            "Type": "Clip"
+        },
+        {
+            "Id": 15440775131389705172,
+            "Type": "Prefab"
+        }
+    ]
```
Comment (JaimeBea) [inline Game/Assets/ybotCasting_Spell.fbx.meta]: There are many unrelated files in the PR that sould be removed.

**Review by JaimeBea — CHANGES_REQUESTED** [review]
You have to upload the .meta files for the test assets. Otherwise they won't work correctly.

File: Project/Source/Components/ComponentParticleSystem.h [hunk 7]
```diff
@@ -68,11 +68,12 @@ class ComponentParticleSystem : public Component {
 	bool alphaTransparency = false; // Enables Alpha Transparency of the image and the color
 	bool isRandomFrame = false;
 	bool randomDirection = false;
+	bool sizeOverTime = false;
```
[the remark this answers is not in the document]
Comment (telmiuse) [reply Project/Source/Components/ComponentParticleSystem.h:71]: not yet , ill ad a comment TODO

[the remark this answers is not in the document]
Comment (telmiuse) [reply Project/Source/Components/ComponentParticleSystem.h:71]: Actually, im almost done for another pr with particles, and this var is using

File: Project/Source/Components/ComponentTrail.cpp (hunk 2 above)
Comment (telmiuse) [reply Project/Source/Components/ComponentTrail.cpp]: it alredy deleted XD

**Review by JaimeBea — APPROVED** [review]

**Comment by JaimeBea** [comment]
Remember to remove test assets before merging

File: Project/Source/Components/ComponentTrail.cpp [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	glDepthMask(GL_FALSE);
+	glEnable(GL_BLEND);
+	glBlendEquation(GL_FUNC_ADD);
+	glBlendFunc(GL_ONE, GL_ONE);
+	glDisable(GL_CULL_FACE);
+
+	unsigned int quadVBO;
+	glGenBuffers(1, &quadVBO);
```
Comment (Jordi-Gil) [inline Project/Source/Components/ComponentTrail.cpp:161]: No need to generate the VBO each frame. Cause the Vertex attributes are always in the same location. For now, you can set the quadVBO as a static variable to check if it's different of 0 to avoid the glGenBuffers each frame.

File: Game/Assets/Shaders/Trail.glsl [hunk 9]
```diff
@@ -0,0 +1,33 @@
+--- vertex
+layout(location = 0) in vec3 vertexPosition;
+layout(location = 1) in vec2 vertexUV0;
+
+uniform mat4 proj;
+uniform mat4 view;
+uniform mat4 model;
```
Comment (Jordi-Gil) [inline Game/Assets/Shaders/Trail.glsl:7]: Unused uniform

File: Game/Assets/Shaders/DrawInstance.glsl [hunk 10]
```diff
@@ -0,0 +1,61 @@
+--- vertex
+
+layout(location = 0) in vec3 vertexPosition;
+layout(location = 1) in vec2 vertexUV0;
+layout(location = 3) in vec2 instanceMatrix;
```
Comment (Jordi-Gil) [inline Game/Assets/Shaders/DrawInstance.glsl:5]: Why location 3?

**Review by FrancPS — APPROVED** [review]
I can't see the trails appearing in GAME playing Trail.scene.
Probably it's me that I have something setted up incorrecty for the configuration that I'm forgetting.

Funcionality looks fine.

Not sure if the serialisation is complete, in case you want to store the trail settings for any reason? Mark it down for a future update if that is the case.

File: Project/Source/Components/ComponentTrail.cpp (hunk 8 above)
Comment (telmiuse) [reply Project/Source/Components/ComponentTrail.cpp:161]: This quad is generated each frame, that is why is there

**Review by Jordi-Gil — APPROVED** [review]
